src/scratch_comment_icon.js

```diff
@@ -0,0 +1,221 @@
+    let left;
+    let right;
+    if (this.sourceBlock.RTL) {
+      left = sourceBlockOrigin.x - this.sourceBlock.width;
+      right = sourceBlockOrigin.x;
+    } else {
+      left = sourceBlockOrigin.x;
+      right = sourceBlockOrigin.x + this.sourceBlock.width;
+    }
+    const blockRect = new Blockly.utils.Rect(
+        sourceBlockOrigin.y, sourceBlockOrigin.y + this.sourceBlock.height, left, right);
+    const y = blockRect.top + this.offsetInBlock.y;
+    const x = this.sourceBlock.workspace.RTL ? blockRect.left : blockRect.right;
+    return new Blockly.utils.Coordinate(x, y);
+  }
+
+  onLocationChange(blockOrigin) {
+    const initialLocation = this.workspaceLocation;
+    super.onLocationChange(blockOrigin);
+    const newLocation = this.workspaceLocation;
+
+    if (this.commentBubble) {
+      const oldBubbleLocation = this.commentBubble.getRelativeToSurfaceXY();
+      const delta = Blockly.utils.Coordinate.difference(newLocation, initialLocation);
+      const newBubbleLocation = Blockly.utils.Coordinate.sum(oldBubbleLocation, delta);
+      this.commentBubble.moveTo(newBubbleLocation);
+      this.commentBubble.setAnchorLocation(this.getAnchorPoint());
+      Blockly.Events.fire(
+        new (Blockly.Events.get('block_comment_move'))(
+          this.commentBubble,
+          oldBubbleLocation,
+          newBubbleLocation,
+        ),
+      );
+    }
+  }
+
+  setText(text) {
+    this.commentBubble?.setText(text);
+  }
+
+  getText() {
+    return this.commentBubble?.getText() ?? '';
+  }
+
+  onTextChanged(oldText, newText) {
+    Blockly.Events.fire(
+      new (Blockly.Events.get(Blockly.Events.BLOCK_CHANGE))(
+        this.sourceBlock,
+        'comment',
+        null,
+        oldText,
+        newText,
+      ),
+    );
+    Blockly.Events.fire(
+      new (Blockly.Events.get('block_comment_change'))(
+        this.commentBubble,
+        oldText,
+        newText,
+      ),
+    );
+  }
+
+  onCollapsed(collapsed) {
+    Blockly.Events.fire(
+      new (Blockly.Events.get('block_comment_collapse'))(
+        this.commentBubble,
+        collapsed,
+      ),
+    );
+  }
+
+  onSizeChanged(oldSize, newSize) {
+    Blockly.Events.fire(
+      new (Blockly.Events.get('block_comment_resize'))(
+        this.commentBubble,
+        oldSize,
+        newSize,
+      ),
+    );
+  }
+
+  setBubbleSize(size) {
+    this.commentBubble?.setSize(size);
+  }
+
+  getBubbleSize() {
+    return this.commentBubble?.getSize() ?? new Blockly.utils.Size(0, 0);
+  }
+
+  setBubbleLocation(newLocation) {
+    const oldLocation = this.getBubbleLocation();
+    this.commentBubble?.moveTo(newLocation);
+    Blockly.Events.fire(
+      new (Blockly.Events.get('block_comment_move'))(
+        this.commentBubble,
+        oldLocation,
+        newLocation,
+      ),
+    );
+  }
+
+  getBubbleLocation() {
+    return this.commentBubble?.getRelativeToSurfaceXY();
+  }
+
+  saveState() {
+    if (this.commentBubble) {
```

Comment:
Can you use an early return here instead?

Reply:
Done